File: engine/dynamic/render_vulkan/src/renderer/bucket_proc.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/debug.hpp"

#include "argus/render/defines.h"

#include "internal/render_vulkan/defines.hpp"
#include "internal/render_vulkan/renderer/bucket_proc.hpp"
#include "internal/render_vulkan/state/render_bucket.hpp"
#include "internal/render_vulkan/state/renderer_state.hpp"
#include "internal/render_vulkan/state/scene_state.hpp"
#include "internal/render_vulkan/util/buffer.hpp"
#include "internal/render_vulkan/util/memory.hpp"

#include <climits>
#include <cstddef>
#include <cstdint>

namespace argus {
    static void _try_free_buffer(BufferInfo &buffer) {
        if (buffer.handle != nullptr) {
            free_buffer(buffer);
            buffer.handle = nullptr;
        }
    }

    void fill_buckets(SceneState &scene_state) {
        const auto &state = scene_state.parent_state;

        for (auto it = scene_state.render_buckets.begin(); it != scene_state.render_buckets.end();) {
            auto *bucket = it->second;

            if (bucket->ubo_buffer.handle == VK_NULL_HANDLE) {
                bucket->ubo_buffer = alloc_buffer(state.device, SHADER_UBO_OBJ_LEN,
                        VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT, GraphicsMemoryPropCombos::DeviceRw);

                // we assume these values will never change

                float uv_stride[] = { bucket->atlas_stride.x, bucket->atlas_stride.y };
                write_to_buffer(bucket->ubo_buffer, uv_stride, SHADER_UNIFORM_OBJ_UV_STRIDE_OFF, sizeof(uv_stride));

                write_val_to_buffer(bucket->ubo_buffer, bucket->light_opacity, SHADER_UNIFORM_OBJ_LIGHT_OPACITY_OFF);
            }

            if (bucket->objects.empty()) {
                _try_free_buffer(it->second->vertex_buffer);
                _try_free_buffer(it->second->anim_frame_buffer);
                _try_free_buffer(it->second->staging_vertex_buffer);
                _try_free_buffer(it->second->staging_anim_frame_buffer);
                _try_free_buffer(it->second->ubo_buffer);
                it->second->destroy();

                it = scene_state.render_buckets.erase(it);

                continue;
            }

            auto pipeline_it = scene_state.parent_state.material_pipelines.find(bucket->material_res.prototype.uid);
            affirm_precond(pipeline_it != scene_state.parent_state.material_pipelines.cend(),
                    "Cannot find material pipeline");

            // the pipeline should have been built during object processing
            auto &pipeline = scene_state.parent_state.material_pipelines
                    .find(bucket->material_res.prototype.uid)->second;

            auto attr_position_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_POSITION);
            auto attr_normal_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_NORMAL);
            auto attr_color_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_COLOR);
            auto attr_texcoord_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_TEXCOORD);
            //auto attr_anim_frame_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_ANIM_FRAME);

            uint32_t vertex_comps = (attr_position_loc.has_value() ? SHADER_ATTRIB_POSITION_LEN : 0)
                    + (attr_normal_loc.has_value() ? SHADER_ATTRIB_NORMAL_LEN : 0)
                    + (attr_color_loc.has_value() ? SHADER_ATTRIB_COLOR_LEN : 0)
                    + (attr_texcoord_loc.has_value() ? SHADER_ATTRIB_TEXCOORD_LEN : 0);

            size_t anim_frame_buf_len = 0;
            if (bucket->needs_rebuild) {
                size_t vertex_buf_len = 0;
                for (auto &obj : bucket->objects) {
                    vertex_buf_len += obj->staging_buffer.size;
                    anim_frame_buf_len += obj->vertex_count * SHADER_ATTRIB_ANIM_FRAME_LEN * sizeof(float);
                }

                _try_free_buffer(bucket->vertex_buffer);
                _try_free_buffer(bucket->anim_frame_buffer);
                _try_free_buffer(bucket->staging_vertex_buffer);
                _try_free_buffer(bucket->staging_anim_frame_buffer);

                affirm_precond(vertex_buf_len <= INT_MAX, "Buffer length is too big");

                bucket->vertex_buffer = alloc_buffer(state.device, vertex_buf_len,
                        VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_VERTEX_BUFFER_BIT,
                        GraphicsMemoryPropCombos::DeviceRo);
                bucket->staging_vertex_buffer = alloc_buffer(state.device, vertex_buf_len,
                        VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT
                                | VK_BUFFER_USAGE_VERTEX_BUFFER_BIT,
                        GraphicsMemoryPropCombos::DeviceRo);

                auto stride = vertex_comps * uint32_t(sizeof(float));
                affirm_precond(stride <= INT_MAX, "Vertex stride is too big");

                affirm_precond(anim_frame_buf_len <= INT_MAX, "Animation frame buffer length is too big");
                bucket->anim_frame_buffer = alloc_buffer(state.device, anim_frame_buf_len,
                        VK_BUFFER_USAGE_TRANSFER_DST_BIT | VK_BUFFER_USAGE_VERTEX_BUFFER_BIT,
                        GraphicsMemoryPropCombos::DeviceRo);
                bucket->staging_anim_frame_buffer = alloc_buffer(state.device, anim_frame_buf_len,
                        VK_BUFFER_USAGE_TRANSFER_SRC_BIT | VK_BUFFER_USAGE_VERTEX_BUFFER_BIT,
                        GraphicsMemoryPropCombos::DeviceRw);
            } else {
                anim_frame_buf_len = bucket->vertex_count * SHADER_ATTRIB_ANIM_FRAME_LEN * sizeof(float);
            }

            bucket->vertex_count = 0;

            bool anim_buf_updated = false;

            size_t offset = 0;
            size_t anim_frame_off = 0;

            float *anim_frame_buf = nullptr;

            for (auto *processed : bucket->objects) {
                if (processed == nullptr) {
                    continue;
                }

                if (bucket->needs_rebuild || processed->updated) {
                    affirm_precond(offset <= INT_MAX, "Buffer offset is too big");
                    affirm_precond(processed->staging_buffer.size <= INT_MAX, "Buffer offset is too big");

                    copy_buffer(state.copy_cmd_buf[state.cur_frame], processed->staging_buffer, 0,
                            bucket->staging_vertex_buffer, offset, processed->staging_buffer.size);

                    processed->updated = false;
                }

                if (bucket->needs_rebuild || processed->anim_frame_updated) {
                    if (anim_frame_buf == nullptr) {
                        anim_frame_buf = reinterpret_cast<float *>(bucket->staging_anim_frame_buffer.mapped);
                    }

                    for (size_t i = 0; i < processed->vertex_count; i++) {
                        anim_frame_buf[anim_frame_off++] = float(processed->anim_frame.x);
                        anim_frame_buf[anim_frame_off++] = float(processed->anim_frame.y);
                    }
                    processed->anim_frame_updated = false;
                    anim_buf_updated = true;
                } else {
                    anim_frame_off += processed->vertex_count * SHADER_ATTRIB_ANIM_FRAME_LEN;
                }

                offset += processed->staging_buffer.size;

                bucket->vertex_count += processed->vertex_count;
            }

            copy_buffer(state.copy_cmd_buf[state.cur_frame], bucket->staging_vertex_buffer, 0, bucket->vertex_buffer, 0,
                    bucket->staging_vertex_buffer.size);
            if (anim_buf_updated) {
                affirm_precond(anim_frame_buf_len <= INT_MAX, "Animated frame buffer length is too big");
                copy_buffer(state.copy_cmd_buf[state.cur_frame], bucket->staging_anim_frame_buffer, 0,
                        bucket->anim_frame_buffer, 0, anim_frame_buf_len);
            }

            bucket->needs_rebuild = false;

            it++;
        }
    }
}

File: engine/modules/render_vulkan/src/renderer/bucket_proc.rs
use std::collections::HashMap;
use ash::vk;
use argus_render::constants::*;
use argus_resman::ResourceIdentifier;
use crate::setup::device::VulkanDevice;
use crate::setup::instance::VulkanInstance;
use crate::state::Scene2dState;
use crate::util::{CommandBufferInfo, PipelineInfo, VulkanBuffer, MEM_CLASS_DEVICE_RO, MEM_CLASS_DEVICE_RW};
use crate::util::defines::*;

pub(crate) fn fill_buckets(
    instance: &VulkanInstance,
    device: &VulkanDevice,
    scene_state: &mut Scene2dState,
    copy_cmd_buf: &CommandBufferInfo,
    material_pipelines: &HashMap<ResourceIdentifier, PipelineInfo>,
) {
    scene_state.render_buckets.retain(|_, bucket| !bucket.objects.is_empty());

    for (_, bucket) in &mut scene_state.render_buckets {
        if bucket.ubo_buffer.is_none() {
            bucket.ubo_buffer = Some(VulkanBuffer::new(
                instance,
                device,
                SHADER_UBO_OBJ_LEN as vk::DeviceSize,
                vk::BufferUsageFlags::UNIFORM_BUFFER,
                MEM_CLASS_DEVICE_RW,
            ).unwrap());
            let ubo_buffer = bucket.ubo_buffer.as_mut().unwrap();

            // we assume these values will never change

            let uv_stride = [bucket.atlas_stride.x, bucket.atlas_stride.y];
            ubo_buffer.write(
                device,
                &uv_stride,
                SHADER_UNIFORM_OBJ_UV_STRIDE_OFF as vk::DeviceSize,
            ).unwrap();

            ubo_buffer.write(
                device,
                &[bucket.light_opacity],
                SHADER_UNIFORM_OBJ_LIGHT_OPACITY_OFF as vk::DeviceSize,
            ).unwrap();
        }

        let pipeline = material_pipelines.get(&bucket.material_res.get_prototype().uid)
            .expect("Cannot find material pipeline");

        let attr_position_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_POSITION);
        let attr_normal_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_NORMAL);
        let attr_color_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_COLOR);
        let attr_texcoord_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_TEXCOORD);
        //let attr_anim_frame_loc = pipeline.reflection.get_attr_loc(SHADER_ATTRIB_ANIM_FRAME);

        let vertex_comps = attr_position_loc.map(|_| SHADER_ATTRIB_POSITION_LEN).unwrap_or(0)
            + attr_normal_loc.map(|_| SHADER_ATTRIB_NORMAL_LEN).unwrap_or(0)
            + attr_color_loc.map(|_| SHADER_ATTRIB_COLOR_LEN).unwrap_or(0)
            + attr_texcoord_loc.map(|_| SHADER_ATTRIB_TEXCOORD_LEN).unwrap_or(0);

        let mut anim_frame_buf_len: vk::DeviceSize = 0;
        if bucket.needs_rebuild {
            let mut vertex_buf_len: vk::DeviceSize = 0;
            for obj_handle in &bucket.objects {
                let obj = &scene_state.processed_objs[obj_handle];
                
                vertex_buf_len += obj.staging_buffer.as_ref().unwrap().len();
                anim_frame_buf_len += (
                    obj.vertex_count *
                        SHADER_ATTRIB_ANIM_FRAME_LEN *
                        size_of::<f32>() as u32
                ) as vk::DeviceSize;
            }

            assert!(vertex_buf_len <= i32::MAX as vk::DeviceSize, "Buffer length is too big");

            if let Some(buf) = bucket.vertex_buffer.take() {
                buf.destroy(device);
            }
            bucket.vertex_buffer.replace(VulkanBuffer::new(
                instance,
                device,
                vertex_buf_len,
                {
                    vk::BufferUsageFlags::TRANSFER_DST |
                    vk::BufferUsageFlags::VERTEX_BUFFER
                },
                MEM_CLASS_DEVICE_RO,
            ).unwrap());
            if let Some(buf) = bucket.staging_vertex_buffer.take() {
                buf.destroy(device);
            }
            bucket.staging_vertex_buffer.replace(VulkanBuffer::new(
                instance,
                device,
                vertex_buf_len,
                {
                    vk::BufferUsageFlags::TRANSFER_SRC |
                    vk::BufferUsageFlags::TRANSFER_DST |
                    vk::BufferUsageFlags::VERTEX_BUFFER
                },
                MEM_CLASS_DEVICE_RO,
            ).unwrap());

            let stride = vertex_comps * size_of::<f32>() as u32;
            assert!(stride <= i32::MAX as u32, "Vertex stride is too big");

            assert!(
                anim_frame_buf_len <= i32::MAX as vk::DeviceSize,
                "Animation frame buffer length is too big",
            );
            if let Some(buf) = bucket.anim_frame_buffer.take() {
                buf.destroy(device);
            }
            bucket.anim_frame_buffer.replace(VulkanBuffer::new(
                instance,
                device,
                anim_frame_buf_len,
                {
                    vk::BufferUsageFlags::TRANSFER_DST |
                    vk::BufferUsageFlags::VERTEX_BUFFER
                },
                MEM_CLASS_DEVICE_RO,
            ).unwrap());
            if let Some(buf) = bucket.staging_anim_frame_buffer.take() {
                buf.destroy(device);
            }
            bucket.staging_anim_frame_buffer.replace(VulkanBuffer::new(
                instance,
                device,
                anim_frame_buf_len,
                {
                    vk::BufferUsageFlags::TRANSFER_SRC |
                    vk::BufferUsageFlags::VERTEX_BUFFER
                },
                MEM_CLASS_DEVICE_RW,
            ).unwrap());
        } else {
            anim_frame_buf_len = (
                bucket.vertex_count *
                    SHADER_ATTRIB_ANIM_FRAME_LEN *
                    size_of::<f32>() as u32
            ) as vk::DeviceSize;
        }

        bucket.vertex_count = 0;

        let mut anim_buf_updated = false;

        let mut offset: vk::DeviceSize = 0;
        let mut anim_frame_off: usize = 0;

        let mut anim_frame_buf = bucket.staging_anim_frame_buffer.as_mut().unwrap().map(
            device,
            0,
            vk::WHOLE_SIZE,
            vk::MemoryMapFlags::empty(),
        )
            .unwrap();
        let anim_frame_buf_map = anim_frame_buf.as_slice_mut::<f32>();

        for obj_handle in &mut bucket.objects {
            let obj = scene_state.processed_objs.get_mut(obj_handle).unwrap();
            
            let staging_buffer = obj.staging_buffer.as_mut().unwrap();

            if bucket.needs_rebuild || obj.updated {
                assert!(
                    staging_buffer.len() <= usize::MAX as vk::DeviceSize,
                    "Buffer offset is too big",
                );

                bucket.staging_vertex_buffer.as_mut().unwrap().copy_from(
                    device,
                    &staging_buffer,
                    0, // src offset
                    copy_cmd_buf,
                    offset, // dst offset
                    vk::WHOLE_SIZE,
                );

                obj.updated = false;
            }

            if bucket.needs_rebuild || obj.anim_frame_updated {                
                for _ in 0..obj.vertex_count {
                    anim_frame_buf_map[anim_frame_off] = obj.anim_frame.x as f32;
                    anim_frame_buf_map[anim_frame_off + 1] = obj.anim_frame.y as f32;
                    anim_frame_off += 2;
                }
                obj.anim_frame_updated = false;
                anim_buf_updated = true;
            } else {
                anim_frame_off += (obj.vertex_count * SHADER_ATTRIB_ANIM_FRAME_LEN) as usize;
            }

            offset += staging_buffer.len();

            bucket.vertex_count += obj.vertex_count;
        }
        drop(anim_frame_buf);

        bucket.vertex_buffer.as_mut().unwrap().copy_from(
            device,
            bucket.staging_vertex_buffer.as_ref().unwrap(),
            0, // src offset
            copy_cmd_buf,
            0, // dst offset
            vk::WHOLE_SIZE,
        );
        if anim_buf_updated {
            assert!(
                anim_frame_buf_len <= i32::MAX as vk::DeviceSize,
                "Animated frame buffer length is too big",
            );
            bucket.anim_frame_buffer.as_mut().unwrap().copy_from(
                device,
                bucket.staging_anim_frame_buffer.as_ref().unwrap(),
                0, // src offset
                copy_cmd_buf,
                0, // dst offset
                anim_frame_buf_len,
            );
        }

        bucket.needs_rebuild = false;
    }
}